### ISSUE TITLE
Unable to start mount-s3 when no IMDS available (timeout/hang)

### ISSUE THREAD

**Issue by dannycjones** [awslabs/mountpoint-s3#571]
### Mountpoint for Amazon S3 version

mount-s3 1.0.2

### AWS Region

eu-west-1

### Describe the running environment

Running in a container. It only appears to show up on old version of Ubuntu as the host OS.

See below some commands: 

```
docker pull public.ecr.aws/ubuntu/ubuntu:20.04
docker run -it --rm --cap-add SYS_ADMIN --device /dev/fuse \
    --env ARCH --env VERSION \
    public.ecr.aws/ubuntu/ubuntu:20.04 /bin/bash

<SNIP> INSTALL MOUNTPOINT-S3 </SNIP>

mount-s3 <BUCKET> ~/mnt
```

```
aws ec2 --region eu-west-1 modify-instance-metadata-options --instance-id i-06d7405b1a250b99b
{
    "InstanceId": "i-06d7405b1a250b99b",
    "InstanceMetadataOptions": {
        "State": "pending",
        "HttpTokens": "required",
        "HttpPutResponseHopLimit": 1,
        "HttpEndpoint": "enabled"
    }
}
```

EC2 Instance Type: `m6i.4xlarge`
AMI: ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20211129 (`ami-08ca3fed11864d6bb`)

### What happened?

When starting in background mode, the process exists after waiting 30 seconds for the mount to complete.

When starting in foreground mode, the process hangs (at least five minutes) until it is killed by the user.

**Comment by dannycjones** [comment]
If you are experiencing this issue, you can turn off the IMDS lookup:

```
AWS_EC2_METADATA_DISABLED=true
```

In this issue, we should fix the timeouts so that IMDS lookups fail gracefully.

**Comment by jamesbornholt** [comment]
Seems like the IMDS client gets stuck in an infinite retry loop when the `PUT` for a token fails:
```
root@443a9c73b5bb:~# grep -E '(GET|PUT) ' /data/mountpoint-s3-2023-10-23T15-51-53Z.log
2023-10-23T15:51:53.624407Z DEBUG awscrt::http-stream: id=0x7fdb88012b10: Created client request on connection=0x7fdb880122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:51:56.639880Z DEBUG awscrt::http-stream: id=0x7fdb6c012b10: Created client request on connection=0x7fdb6c0122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:51:59.640935Z DEBUG awscrt::http-stream: id=0x7fdb74012a90: Created client request on connection=0x7fdb740122e0: GET /latest/dynamic/instance-identity/document HTTP/1.1
2023-10-23T15:51:59.641921Z DEBUG awscrt::http-stream: id=0x7fdb64014750: Created client request on connection=0x7fdb64013f30: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:02.657422Z DEBUG awscrt::http-stream: id=0x7fdb840137a0: Created client request on connection=0x7fdb84013090: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:05.659281Z DEBUG awscrt::http-stream: id=0x7fdb6c012b10: Created client request on connection=0x7fdb6c0122e0: GET /latest/dynamic/instance-identity/document HTTP/1.1
2023-10-23T15:52:05.661371Z DEBUG awscrt::http-stream: id=0x7fdb7c012b10: Created client request on connection=0x7fdb7c0122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:08.676363Z DEBUG awscrt::http-stream: id=0x7fdb7c012b10: Created client request on connection=0x7fdb7c0122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:11.677324Z DEBUG awscrt::http-stream: id=0x7fdb64014750: Created client request on connection=0x7fdb64013f30: GET /latest/dynamic/instance-identity/document HTTP/1.1
2023-10-23T15:52:11.678370Z DEBUG awscrt::http-stream: id=0x7fdb88012b10: Created client request on connection=0x7fdb880122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:14.694672Z DEBUG awscrt::http-stream: id=0x7fdb64014750: Created client request on connection=0x7fdb64013f30: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:17.695626Z DEBUG awscrt::http-stream: id=0x7fdb74012a90: Created client request on connection=0x7fdb740122e0: GET /latest/dynamic/instance-identity/document HTTP/1.1
2023-10-23T15:52:17.696634Z DEBUG awscrt::http-stream: id=0x7fdb74012a90: Created client request on connection=0x7fdb740122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:20.717507Z DEBUG awscrt::http-stream: id=0x7fdb840137a0: Created client request on connection=0x7fdb84013090: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:23.718481Z DEBUG awscrt::http-stream: id=0x7fdb7c012b10: Created client request on connection=0x7fdb7c0122e0: GET /latest/dynamic/instance-identity/document HTTP/1.1
2023-10-23T15:52:23.719457Z DEBUG awscrt::http-stream: id=0x7fdb88012b10: Created client request on connection=0x7fdb880122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:26.743396Z DEBUG awscrt::http-stream: id=0x7fdb6c012b10: Created client request on connection=0x7fdb6c0122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:29.744473Z DEBUG awscrt::http-stream: id=0x7fdb64014750: Created client request on connection=0x7fdb64013f30: GET /latest/dynamic/instance-identity/document HTTP/1.1
2023-10-23T15:52:29.745869Z DEBUG awscrt::http-stream: id=0x7fdb80014310: Created client request on connection=0x7fdb80013c60: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:32.761099Z DEBUG awscrt::http-stream: id=0x7fdb74012a90: Created client request on connection=0x7fdb740122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:35.762050Z DEBUG awscrt::http-stream: id=0x7fdb88012b10: Created client request on connection=0x7fdb880122e0: GET /latest/dynamic/instance-identity/document HTTP/1.1
2023-10-23T15:52:35.763298Z DEBUG awscrt::http-stream: id=0x7fdb80014310: Created client request on connection=0x7fdb80013c60: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:38.767964Z DEBUG awscrt::http-stream: id=0x7fdb7c012b10: Created client request on connection=0x7fdb7c0122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:41.768900Z DEBUG awscrt::http-stream: id=0x7fdb78013670: Created client request on connection=0x7fdb78013060: GET /latest/dynamic/instance-identity/document HTTP/1.1
2023-10-23T15:52:41.770007Z DEBUG awscrt::http-stream: id=0x7fdb840137a0: Created client request on connection=0x7fdb84013090: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:44.788396Z DEBUG awscrt::http-stream: id=0x7fdb64014750: Created client request on connection=0x7fdb64013f30: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:47.789491Z DEBUG awscrt::http-stream: id=0x7fdb6c012b10: Created client request on connection=0x7fdb6c0122e0: GET /latest/dynamic/instance-identity/document HTTP/1.1
2023-10-23T15:52:47.790616Z DEBUG awscrt::http-stream: id=0x7fdb7c012b10: Created client request on connection=0x7fdb7c0122e0: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:50.813517Z DEBUG awscrt::http-stream: id=0x7fdb840137a0: Created client request on connection=0x7fdb84013090: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:53.814574Z DEBUG awscrt::http-stream: id=0x7fdb840137a0: Created client request on connection=0x7fdb84013090: GET /latest/dynamic/instance-identity/document HTTP/1.1
2023-10-23T15:52:53.815627Z DEBUG awscrt::http-stream: id=0x7fdb64014750: Created client request on connection=0x7fdb64013f30: PUT /latest/api/token HTTP/1.1
2023-10-23T15:52:56.834250Z DEBUG awscrt::http-stream: id=0x7fdb78013670: Created client request on connection=0x7fdb78013060: PUT /latest/api/token HTTP/1.1
```
Wild guess: I wonder if the nested "get a token first, then do the actual request" is sharing a retry strategy across both requests and getting confused.

**Comment by jamesbornholt** [comment]
Reproduced on m6i.2xlarge running Ubuntu 20.04, but I had to change the IMDS settings for the instance to enforce IMDSv2 first (and also confirm that it was already configured with a 1-hop limit):
```
[15:36:03] ~/c/mountpoint-s3 (main|✔) $ aws ec2 modify-instance-metadata-options --instance-id i-00e26372c49458ebb --http-tokens required
{
    "InstanceId": "i-00e26372c49458ebb",
    "InstanceMetadataOptions": {
        "State": "pending",
        "HttpTokens": "required",
        "HttpPutResponseHopLimit": 1,
        "HttpEndpoint": "enabled",
        "HttpProtocolIpv6": "disabled",
        "InstanceMetadataTags": "disabled"
    }
}
```

**Comment by dannycjones** [comment]
Just tried on the latest commit on `main` and looks OK.

```
bash-5.2# /mount-s3 --version
mount-s3 1.1.0-unofficial+d38d45f
bash-5.2# /mount-s3 djonesoa-dub /mnt
Error: Failed to create S3 client

Caused by:
    0: initial ListObjectsV2 failed for bucket djonesoa-dub in region us-east-1
    1: Client error
    2: No signing credentials found
Error: Failed to create mount process
```

The fix for this will be available in the next release.

**Comment by passaro** [comment]
Fix included in [v1.1.1](https://github.com/awslabs/mountpoint-s3/releases/tag/mountpoint-s3-1.1.1).
